src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -29,9 +29,70 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#define BYTES_TO_U8_SAMPLES	0
 #define BYTES_TO_S16_SAMPLES	1
 #define BYTES_TO_S32_SAMPLES	2
 
+#if CONFIG_PCM_CONVERTER_FORMAT_U8 && CONFIG_PCM_CONVERTER_FORMAT_S32LE
+static int pcm_convert_u8_to_s32(const struct audio_stream __sparse_cache *source,
+				  uint32_t ioffset, struct audio_stream __sparse_cache *sink,
+				  uint32_t ooffset, uint32_t samples)
+{
+	uint8_t *src = audio_stream_get_rptr(source);
+	int32_t *dst = audio_stream_get_wptr(sink);
+	int processed;
+	int nmax, i, n;
+
+	src += ioffset;
+	dst += ooffset;
+	for (processed = 0; processed < samples; processed += n) {
```

Comment:
this might be an example of why I usually use `unsigned` for non-negative loop iterators: `samples` is unsigned, `processed` is incremented by a variable `n`, which is calculated by subtraction, so *purely in theory* if `n` somehow becomes negative and `processed` could become negative too. I'm pretty sure the below logic makes this impossible, but just to be on the safe side in such cases...

Reply:
ok, I changed `int` to `uint32_t` because `samples` is defined as `uint32_t`. However, `audio_stream_bytes_without_wrap` returns `int`, which doesn't fit the puzzle.

Reply:
> audio_stream_bytes_without_wrap returns int, which doesn't fit the puzzle.

In commit a892e0f66cb74 you added an `assert()` to make sure the returned value is always positive. That function should just be fixed to return an unsigned type.

BTW I think at least one of these asserts is buggy because you've used `intptr_t`, see 64bits demo below. Why not just `uintptr_t`?

All this proves once again the point that C integers are an absolute nightmare full of pitfalls (and security vulnerabilities) and that they require extreme care.

```C
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

int main()
{
  void * const ptr1 = (void *)  INT64_MAX;
  void * const ptr2 = (void *) UINT64_MAX ; // -1

  printf("%p, %p\n", ptr1, ptr2);
  printf("%ld, %ld\n", (intptr_t) ptr1, (intptr_t) ptr2);

  printf("ptr1 < ptr2 is %s\n",  (intptr_t) ptr1 <  (intptr_t) ptr2 ? "true" : "false");
  printf("ptr1 < ptr2 is %s\n", (uintptr_t) ptr1 < (uintptr_t) ptr2 ? "true" : "false");

  return 0;
}

ptr1 < ptr2 is false
ptr1 < ptr2 is true
```

Reply:
https://stefansf.de/c-quiz/#integer-constants + many others